pkg/apis/network/v1/egressrouter_types.go

```diff
@@ -0,0 +1,210 @@
+	IP IPConfig `json:"ip"`
+}
+
+// EgressRouterInterface contains the configuration of interface to create/use.
+// +kubebuilder:validation:Required
+type EgressRouterInterface struct {
+	// Arguments specific to the interfaceType macvlan
+	// +kubebuilder:validation:Required
+	MacVLAN MacVLANConfig `json:"macVLAN"`
+}
+
+// MacvlanMode defines the different types of modes that are supported for the macvlan interface.
+// source: https://man7.org/linux/man-pages/man8/ip-link.8.html
+type MacvlanMode string
+
+const (
+	// MacvlanModeBridge connects all endpoints directly to each other, communication is not redirected through the physical interface's peer.
+	MacvlanModeBridge MacvlanMode = "Bridge"
+
+	// MacvlanModePrivate does not allow communication between macvlan instances on the same physical interface,
+	// even if the external switch supports hairpin mode.
+	MacvlanModePrivate MacvlanMode = "Private"
+
+	// MacvlanModeVEPA is the Virtual Ethernet Port Aggregator mode. Data from one macvlan instance to the other on the
+	// same physical interface is transmitted over the physical interface. Either the attached switch needs
+	// to support hairpin mode, or there must be a TCP/IP router forwarding the packets in order to allow
+	// communication. This is the default mode.
+	MacvlanModeVEPA MacvlanMode = "VEPA"
+
+	// MacvlanModePassthru mode gives more power to a single endpoint, usually in macvtap mode.
+	// It is not allowed for more than one endpoint on the same physical interface. All traffic will be forwarded
+	// to this endpoint, allowing virtio guests to change MAC address or set promiscuous mode in order to bridge the
+	// interface or create vlan interfaces on top of it.
+	MacvlanModePassthru MacvlanMode = "Passthru"
+)
+
+// MacVLANConfig consists of arguments specific to the macvlan EgressRouterInterfaceType
+// +kubebuilder:validation:Required
+type MacVLANConfig struct {
+	// Mode depicts the mode that is used for the macvlan interface; one of Bridge|Private|VEPA|Passthru. The default mode is "Bridge".
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=Bridge;Private;VEPA;Passthru
+	// +kubebuilder:default:Bridge
+	Mode MacvlanMode `json:"mode"`
+
+	// Name of the master inteface. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty"`
+}
+
+// IPConfig contains the configuration for the ip arguments
+// +kubebuilder:validation:Required
+type IPConfig struct {
+	// List of IP addresses to configure on the interface.
+	// +kubebuilder:validation:Required
+	Addresses []EgressRouterAddress `json:"addresses"`
+
+	// List of CIDR blocks that the pod is allowed to connect to via this interface. If not provided, the pod can connect to any destination.
+	// +kubebuilder:validation:Optional
+	// +optional
+	AllowedDestinations []CIDR `json:"allowedDestinations,omitempty"`
```

Comment:
I don't think `AllowedDestinations` and `IPConfig` necessarily go together. If we plan to extend this to support the egress router HTTP proxy mode, then we'd need a different kind of `AllowedDestinations` with hostnames for that.

Reply:
hmm I wasn't aware of the HTTP proxy mode design plan, was not included in the enhancement: https://github.com/openshift/enhancements/blob/master/enhancements/network/egress-router.md. The API portion was roughly based on https://github.com/openshift/egress-router-cni/blob/master/README.md keeping in mind the portions daniel said he plans to implement. I don't know if this particular card/epic of writing a CRD needs a separate enhancement in itself, but in any case it should be included in the main enhancement.

Reply:
ok, so we'd have probably either an array of cidrs for the normal router scenario and an array of strings for the proxy mode. hmm I'll try to make this field more generic.